### DISCUSSION TITLE
some questions about ezengine and using it directly in c++ and write games in it

### DISCUSSION THREAD

**Discussion by brightening-eyes** [ezEngine/ezEngine#1285] [Q&A] [answered]
hello.
I want to use ezengine directly as a C++ library and I have some questions:
* first, when compiling it as a submodule, the qt stuff which is for editor gets downloaded always no matter if I set EZ_ENABLE_QT_SUPPORT to OFF or setting a qt directory.
* for representing menus and the game as different states, how can I go for it? I mean changing states. I mean getting from a ezGameState into another
thanks

**Comment by jankrassnigg** [comment]
The Qt thing is a (minor) bug. As I answered on the ticket just now, you could have a look at fixing that.

For UI you can use ImGui or RML UI. ImGui is definitely the easiest to get started with. See here for documentation:
https://ezengine.net/pages/docs/ui/imgui.html

This is also shown in the [RTS](https://ezengine.net/pages/samples/rts.html) sample and the [SampleGamePlugin](https://ezengine.net/pages/samples/sample-game-plugin.html).

**Comment by brightening-eyes** [comment]
hello.
I do not mean UI for now (I know about imgui existance in ezengine).
I mean the states for different menus, the game itself, when the game is paused etc.
I want to switch between game states and have different ezWorld objects.
first, how can I track in which state I am in, and second, How would I change the state?
for example in the main menu, when the user chooses start game or something like that, it should change the world and initialize game objects like player, different enemies and so on and so forth.
or, when the user pauses the game, it should hold the objects in the world (the worlds are different I know I can have up to 64 worlds), and then gets into pause state.
another question:
for different components that do stuff (like custom npc objects), should I implement update() and they get called automatically? I do know for player I need to implement ProcessInputs()
thanks

**Reply by jankrassnigg** [reply, marked as answer]
You can/should only have a single ezGameState. It is basically the game loop for your entire game.

Inside that track the current state of your game (in menu, in game) however you like. Could just be an int variable or could be something much more complex, that's up to you.

You can switch the 'active' ezWorld through some function on ezGameState. The active one will automatically be updated and rendered.
I have an old test project where I implemented dynamic level switching, you can find the code here:
<https://github.com/jankrassnigg/EzTestProject/blob/main/CppSource/EzTestProjectPlugin/GameState/EzTestProjectGameState.cpp>

But this may require some fiddling, the API isn't well documented and there are no official samples that show how to do this, yet.

For custom components, you have to implement [a component manager](https://ezengine.net/pages/docs/runtime/world/component-managers.html). The component manager typically registers a few hooks to get updated every frame. However, the most simple thing you can do, is to use the `ezComponentManagerSimple` class. In that case it will automatically call an `Update` function every frame. It's a good start, and good enough for a lot of components, if you need more than that later, you can adjust your component manager at any time.

For players you CAN implement the logic on the game state, in ProcessInput, but you can also do this inside a player component. See for example the [Monster Attack](https://github.com/ezEngine/project-monster-attack) sample. It depends on your type of game. If you have a proper representation of your player (like in an FPS), this may be better. If you have a top-down game, like an RTS, it's typically easier to do it in the game state.

**Comment by brightening-eyes** [comment]
hello,
thanks for the complete answer.
based on the test project you've put, I think I go with global blackboard approach.
it is more versatile.
I also think message handlers could be a way of changing states (maybe I'm wrong)?
thanks

**Reply by jankrassnigg** [reply]
Messages and definitely a very useful tool in general.
The global blackboards are also very useful. In the end you have multiple ways that you can do things, you have to figure out what makes the most sense for you.
